crates/secrets-core/src/engine.rs

```diff
@@ -0,0 +1,133 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License, Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+use crate::common::ByteSpan;
+use crate::location::PointSpan;
+use crate::matcher::Matcher;
+use crate::rule::{Rule, RuleId};
+use crate::thread_pool::ThreadPool;
+use crate::validator::{Candidate, SecretCategory, Validator};
+use crate::worker::{Worker, WorkerError};
+use std::cell::RefCell;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::ops::{Deref, DerefMut};
+use std::path::{Path, PathBuf};
+use std::rc::Rc;
+use std::sync::{Arc, Mutex};
+
+#[derive(Debug, thiserror::Error)]
+enum EngineError {
+    #[error(transparent)]
+    Io(#[from] io::Error),
+    #[error(transparent)]
+    Worker(#[from] WorkerError),
+}
+
+pub struct Engine<T: ThreadPool> {
+    thread_pool: T,
+    rules: Vec<Rule>,
+    matchers: Vec<Matcher>,
+    validators: Vec<Arc<Mutex<Box<dyn Validator + Send + Sync>>>>,
+    run_validation: bool,
+}
```

Comment:
Why use a custom `ThreadPool` and start to put our finger into concurrency issues when it's a solved problem with rayon and other libraries? Ideally, this should run in the same environment than the static analyzer and this should just be something we can parallelize with rayon. There is a good change that starting our own (untested) concurrency implementation will lead to many bugs down the line.